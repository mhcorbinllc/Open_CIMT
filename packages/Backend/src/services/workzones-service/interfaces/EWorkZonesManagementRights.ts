import {convertDictionaryToViewLabelArray} from "../../../utils/convertDictionaryToViewLabelArray";

export enum EWorkZonesManagementRights {
  WORKZONES_VIEW = "WZM_V",
  WORKZONES_EDIT = "WZM_E",
}

const labelsDic: Record<EWorkZonesManagementRights, string>  = {
  [EWorkZonesManagementRights.WORKZONES_VIEW]: 'CIMs / View',
  [EWorkZonesManagementRights.WORKZONES_EDIT]: 'CIMs / Edit',
};

export const EWorkZonesManagementRightsArray = convertDictionaryToViewLabelArray(labelsDic);
