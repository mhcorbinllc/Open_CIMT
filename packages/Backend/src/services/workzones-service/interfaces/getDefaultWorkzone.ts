import * as moment from "moment";

import {
  EWZBroadcastSelectionMode,
  EWZClosureType,
  EWZLane,
  EWZShoulder,
  IWorkzone,
} from "./IWorkzone";
import {
  EWZClosureSelectionMode,
} from "./IWorkzonesAppSettings";

export const getDefaultWorkzone = (): IWorkzone => {
  const now = moment().seconds(0);

  return {
    workzoneId: '',
    kapschId: -1,
    name: '',
    active: true,

    notes: '',
    start: now.toDate(),
    end: now.clone().add(3, 'days')
      .toDate(),
    forEver: false,

    closureSelectionMode: EWZClosureSelectionMode.SIMPLE,
    closureType: EWZClosureType.NONE,
    itisCodes: [],
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 0,

    closedLaneWidthInMeters: 0,
    workersPresent: false,

    point: {
      lat: 0,
      lng: 0,
      alt: 0,
    },

    path: {points: []},

    heading: [],

    broadcast: {
      selectionMode: EWZBroadcastSelectionMode.REGION,
      selectionByRegion: {region: {points: []}},
      selectionByRadius: {radiusInMeters: 0},
    },

    createdAt: 0,
    updatedAt: 0,
    deletedAt: 0,
  };
};

