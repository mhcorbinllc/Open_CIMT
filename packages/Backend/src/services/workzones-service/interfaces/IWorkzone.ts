import {EWZClosureSelectionMode} from "./IWorkzonesAppSettings";

// Dev note: In case you want to change this interface, consider to update existed data.
// Otherwise the app would break on render!
// - Database, through migration scripts
// - Database, CIM history items
// - Client's indexedDB CIMs items

export interface IWorkzone {
  workzoneId: string; // PK
  kapschId: number;
  active: boolean;

  name: string;
  start: Date;
  end: Date;
  forEver: boolean;

  closureSelectionMode: EWZClosureSelectionMode;
  closureType:
    | EWZClosureType
    | EWZSimpleClosureTypeLane
    | EWZSimpleClosureTypeShoulder
    | EWZSimpleClosureTypeReducedSpeed
    | EWZSimpleClosureTypeLeftTurn
    | EWZSimpleClosureTypeRightTurn
    | EWZSimpleClosureTypeHeightLimit
    | EWZClosureTypeDeprecated
    ;
  itisCodes: number[];
  closedLane: EWZLane;
  closedShoulder: EWZShoulder;
  speedLimitInMiles: number;

  closedLaneWidthInMeters: number;
  workersPresent: boolean;
  notes: string;

  point: IGeoPoint;

  path: IGeoLine;

  heading: EWZHeading[];

  broadcast: ICIMBroadcast;

  createdAt: number;
  updatedAt: number;
  deletedAt: number;
}

export interface ICIMBroadcast {
  selectionMode: EWZBroadcastSelectionMode;
  selectionByRegion: ICIMBroadcastSelectionByRegion;
  selectionByRadius: ICIMBroadcastSelectionByRadius;
}
export interface ICIMBroadcastSelectionByRegion {
  region: IGeoPolygon;
}
export interface ICIMBroadcastSelectionByRadius {
  // Referenced to `point`
  radiusInMeters: number;
}

export enum EWZBroadcastSelectionMode {
  REGION = "REGION",
  RADIUS = "RADIUS",
}

export interface IGeoPoint {
  lat: number;
  lng: number;
  alt?: number;
}

export interface IGeoPolygon {
  points: IGeoPoint[];
}

export interface IGeoLine {
  points: IGeoPoint[];
}

export enum EWZShoulder {
  NONE = "NONE",
  LEFT = "LEFT",
  RIGHT = "RIGHT",
}

export enum EWZLane {
  NONE = "NONE",
  RIGHT = "RIGHT",
  LEFT = "LEFT",
  CENTER = "CENTER",
  ALL = "ALL",
}

export enum EWZSimpleClosureType {
  NONE = "NONE",
  LANE_CLOSED = "LANE_CLOSED",
  SHOULDER_CLOSED = "SHOULDER_CLOSED",
  REDUCED_SPEED = "REDUCED_SPEED",
  LEFT_TURN = "LEFT_TURN",
  RIGHT_TURN = "RIGHT_TURN",
  HEIGHT_LIMIT = "HEIGHT_LIMIT",
}

export enum EWZSimpleClosureTypeLane {
  LEFT_LANE_CLOSED_MERGE_RIGHT = "LEFT_LANE_CLOSED_MERGE_RIGHT",
  RIGHT_LANE_CLOSED_MERGE_LEFT = "RIGHT_LANE_CLOSED_MERGE_LEFT",
}

export enum EWZSimpleClosureTypeShoulder {
  LEFT_SHOULDER_CLOSED = "LEFT_SHOULDER_CLOSED",
  RIGHT_SHOULDER_CLOSED = "RIGHT_SHOULDER_CLOSED",
}

export enum EWZSimpleClosureTypeReducedSpeed {
  REDUCED_SPEED_15_MPH = "REDUCED_SPEED_15_MPH",
  REDUCED_SPEED_20_MPH = "REDUCED_SPEED_20_MPH",
  REDUCED_SPEED_25_MPH = "REDUCED_SPEED_25_MPH",
  REDUCED_SPEED_30_MPH = "REDUCED_SPEED_30_MPH",
  REDUCED_SPEED_35_MPH = "REDUCED_SPEED_35_MPH",
  REDUCED_SPEED_40_MPH = "REDUCED_SPEED_40_MPH",
  REDUCED_SPEED_45_MPH = "REDUCED_SPEED_45_MPH",
  REDUCED_SPEED_50_MPH = "REDUCED_SPEED_50_MPH",
  REDUCED_SPEED_55_MPH = "REDUCED_SPEED_55_MPH",
  REDUCED_SPEED_60_MPH = "REDUCED_SPEED_60_MPH",
  REDUCED_SPEED_65_MPH = "REDUCED_SPEED_65_MPH",
  REDUCED_SPEED_70_MPH = "REDUCED_SPEED_70_MPH",
}

export enum EWZSimpleClosureTypeLeftTurn {
  LEFT_TURN_SPEED_15_MPH = "LEFT_TURN_SPEED_15_MPH",
  LEFT_TURN_SPEED_20_MPH = "LEFT_TURN_SPEED_20_MPH",
  LEFT_TURN_SPEED_25_MPH = "LEFT_TURN_SPEED_25_MPH",
  LEFT_TURN_SPEED_30_MPH = "LEFT_TURN_SPEED_30_MPH",
  LEFT_TURN_SPEED_35_MPH = "LEFT_TURN_SPEED_35_MPH",
  LEFT_TURN_SPEED_40_MPH = "LEFT_TURN_SPEED_40_MPH",
  LEFT_TURN_SPEED_45_MPH = "LEFT_TURN_SPEED_45_MPH",
  LEFT_TURN_SPEED_50_MPH = "LEFT_TURN_SPEED_50_MPH",
  LEFT_TURN_SPEED_55_MPH = "LEFT_TURN_SPEED_55_MPH",
}

export enum EWZSimpleClosureTypeRightTurn {
  RIGHT_TURN_SPEED_15_MPH = "RIGHT_TURN_SPEED_15_MPH",
  RIGHT_TURN_SPEED_20_MPH = "RIGHT_TURN_SPEED_20_MPH",
  RIGHT_TURN_SPEED_25_MPH = "RIGHT_TURN_SPEED_25_MPH",
  RIGHT_TURN_SPEED_30_MPH = "RIGHT_TURN_SPEED_30_MPH",
  RIGHT_TURN_SPEED_35_MPH = "RIGHT_TURN_SPEED_35_MPH",
  RIGHT_TURN_SPEED_40_MPH = "RIGHT_TURN_SPEED_40_MPH",
  RIGHT_TURN_SPEED_45_MPH = "RIGHT_TURN_SPEED_45_MPH",
  RIGHT_TURN_SPEED_50_MPH = "RIGHT_TURN_SPEED_50_MPH",
  RIGHT_TURN_SPEED_55_MPH = "RIGHT_TURN_SPEED_55_MPH",
}

export enum EWZSimpleClosureTypeHeightLimit {
  HEIGHT_LIMIT_14_FT_0_IN = "HEIGHT_LIMIT_14_FT_0_IN",
}

// Using as composite enum for all closure types
export enum EWZClosureType {
  NONE = "NONE",
  LEFT_LANE_CLOSED_MERGE_RIGHT = "LEFT_LANE_CLOSED_MERGE_RIGHT",
  RIGHT_LANE_CLOSED_MERGE_LEFT = "RIGHT_LANE_CLOSED_MERGE_LEFT",
  LEFT_SHOULDER_CLOSED = "LEFT_SHOULDER_CLOSED",
  RIGHT_SHOULDER_CLOSED = "RIGHT_SHOULDER_CLOSED",
  REDUCED_SPEED_15_MPH = "REDUCED_SPEED_15_MPH",
  REDUCED_SPEED_20_MPH = "REDUCED_SPEED_20_MPH",
  REDUCED_SPEED_25_MPH = "REDUCED_SPEED_25_MPH",
  REDUCED_SPEED_30_MPH = "REDUCED_SPEED_30_MPH",
  REDUCED_SPEED_35_MPH = "REDUCED_SPEED_35_MPH",
  REDUCED_SPEED_40_MPH = "REDUCED_SPEED_40_MPH",
  REDUCED_SPEED_45_MPH = "REDUCED_SPEED_45_MPH",
  REDUCED_SPEED_50_MPH = "REDUCED_SPEED_50_MPH",
  REDUCED_SPEED_55_MPH = "REDUCED_SPEED_55_MPH",
  REDUCED_SPEED_60_MPH = "REDUCED_SPEED_60_MPH",
  REDUCED_SPEED_65_MPH = "REDUCED_SPEED_65_MPH",
  REDUCED_SPEED_70_MPH = "REDUCED_SPEED_70_MPH",
  LEFT_TURN_SPEED_15_MPH = "LEFT_TURN_SPEED_15_MPH",
  LEFT_TURN_SPEED_20_MPH = "LEFT_TURN_SPEED_20_MPH",
  LEFT_TURN_SPEED_25_MPH = "LEFT_TURN_SPEED_25_MPH",
  LEFT_TURN_SPEED_30_MPH = "LEFT_TURN_SPEED_30_MPH",
  LEFT_TURN_SPEED_35_MPH = "LEFT_TURN_SPEED_35_MPH",
  LEFT_TURN_SPEED_40_MPH = "LEFT_TURN_SPEED_40_MPH",
  LEFT_TURN_SPEED_45_MPH = "LEFT_TURN_SPEED_45_MPH",
  LEFT_TURN_SPEED_50_MPH = "LEFT_TURN_SPEED_50_MPH",
  LEFT_TURN_SPEED_55_MPH = "LEFT_TURN_SPEED_55_MPH",
  RIGHT_TURN_SPEED_15_MPH = "RIGHT_TURN_SPEED_15_MPH",
  RIGHT_TURN_SPEED_20_MPH = "RIGHT_TURN_SPEED_20_MPH",
  RIGHT_TURN_SPEED_25_MPH = "RIGHT_TURN_SPEED_25_MPH",
  RIGHT_TURN_SPEED_30_MPH = "RIGHT_TURN_SPEED_30_MPH",
  RIGHT_TURN_SPEED_35_MPH = "RIGHT_TURN_SPEED_35_MPH",
  RIGHT_TURN_SPEED_40_MPH = "RIGHT_TURN_SPEED_40_MPH",
  RIGHT_TURN_SPEED_45_MPH = "RIGHT_TURN_SPEED_45_MPH",
  RIGHT_TURN_SPEED_50_MPH = "RIGHT_TURN_SPEED_50_MPH",
  RIGHT_TURN_SPEED_55_MPH = "RIGHT_TURN_SPEED_55_MPH",
  HEIGHT_LIMIT_14_FT_0_IN = "HEIGHT_LIMIT_14_FT_0_IN",
}

export enum EWZClosureTypeDeprecated {
  RIGHT_2_LANES_CLOSED_AHEAD = "RIGHT_2_LANES_CLOSED_AHEAD",
  LEFT_2_LANES_CLOSED_AHEAD = "LEFT_2_LANES_CLOSED_AHEAD",
  LEFT_LANE_CLOSED_AHEAD_MERGE_RIGHT = "LEFT_LANE_CLOSED_AHEAD_MERGE_RIGHT",
  RIGHT_LANE_CLOSED_AHEAD_MERGE_LEFT = "RIGHT_LANE_CLOSED_AHEAD_MERGE_LEFT",
  LEFT_SHOULDER_CLOSED_AHEAD = "LEFT_SHOULDER_CLOSED_AHEAD",
  RIGHT_SHOULDER_CLOSED_AHEAD = "RIGHT_SHOULDER_CLOSED_AHEAD",
  REDUCED_SPEED_AHEAD_15_MPH = "REDUCED_SPEED_AHEAD_15_MPH",
  REDUCED_SPEED_AHEAD_20_MPH = "REDUCED_SPEED_AHEAD_20_MPH",
  REDUCED_SPEED_AHEAD_25_MPH = "REDUCED_SPEED_AHEAD_25_MPH",
  REDUCED_SPEED_AHEAD_30_MPH = "REDUCED_SPEED_AHEAD_30_MPH",
  REDUCED_SPEED_AHEAD_35_MPH = "REDUCED_SPEED_AHEAD_35_MPH",
  REDUCED_SPEED_AHEAD_40_MPH = "REDUCED_SPEED_AHEAD_40_MPH",
  REDUCED_SPEED_AHEAD_45_MPH = "REDUCED_SPEED_AHEAD_45_MPH",
  REDUCED_SPEED_AHEAD_50_MPH = "REDUCED_SPEED_AHEAD_50_MPH",
  REDUCED_SPEED_AHEAD_55_MPH = "REDUCED_SPEED_AHEAD_55_MPH",
  REDUCED_SPEED_AHEAD_60_MPH = "REDUCED_SPEED_AHEAD_60_MPH",
  REDUCED_SPEED_AHEAD_65_MPH = "REDUCED_SPEED_AHEAD_65_MPH",
  REDUCED_SPEED_AHEAD_70_MPH = "REDUCED_SPEED_AHEAD_70_MPH",
}

export enum EWZHeading {
    NORTH = "from000-0to022-5degrees",
    NORTH_NORTH_EAST = "from022-5to045-0degrees",
    NORTH_EAST = "from045-0to067-5degrees",
    EAST_NORTH_EAST = "from067-5to090-0degrees",
    EAST = "from090-0to112-5degrees",
    EAST_SOUTH_EAST = "from112-5to135-0degrees",
    SOUTH_EAST = "from135-0to157-5degrees",
    SOUTH_SOUTH_EAST = "from157-5to180-0degrees",
    SOUTH = "from180-0to202-5degrees",
    SOUTH_SOUTH_WEST = "from202-5to225-0degrees",
    SOUTH_WEST = "from225-0to247-5degrees",
    WEST_SOUTH_WEST = "from247-5to270-0degrees",
    WEST = "from270-0to292-5degrees",
    WEST_NORTH_WEST = "from292-5to315-0degrees",
    NORTH_WEST = "from315-0to337-5degrees",
    NORTH_NORTH_WEST = "from337-5to360-0degrees",
}
