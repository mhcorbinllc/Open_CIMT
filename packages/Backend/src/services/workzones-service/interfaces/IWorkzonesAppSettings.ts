import {IWorkzone} from "./IWorkzone";
import {getDefaultAppItisCodes} from "./getDefaultAppItisCodes";
import {closureTypePresetValues} from "./closureTypePresetValues";
// App settings per company (not per device or user).

export interface IWorkzonesAppSettings {
  // See the apiWorkzonesAppSettingPost.ts's validation for the valid options per property

  deviceId: string;           // PK, "*" is for all devices of the company

  serviceChannel: string;
  psid: string;               // Courier font type
  txmode: EWZTXMode;
  intervalInMs: number;
  priority: number;           // 0..63
  useSignature: boolean;
  useEncryption: boolean;
  closureSelectionMode: EWZClosureSelectionMode;

  itisCodes: IItisCode[];

  createdAt: number;
  updatedAt: number;
  deletedAt: number;
}

export enum EWZTXMode {
  CONT = "CONT",
  ALT = "ALT",
}

export enum EWZClosureSelectionMode {
  CUSTOM = "CUSTOM",
  SIMPLE = "SIMPLE",
}

export interface IItisCode {
  code: number;
  label: string;
}

export const defaultWorkzoneAppSettings: IWorkzonesAppSettings = {
  deviceId: '',
  serviceChannel: "172",
  psid: '8002',
  txmode: EWZTXMode.CONT,
  intervalInMs: 1000,
  priority: 4,
  useSignature: false,
  useEncryption: false,
  closureSelectionMode: EWZClosureSelectionMode.SIMPLE,
  itisCodes: getDefaultAppItisCodes(),
  createdAt: 0,
  updatedAt: 0,
  deletedAt: 0,
};

export interface IValidationErrorItisRemoveBlockedByWorkzone {
  removedItisCodes: number[];
  workzone: IWorkzone;
}

export interface IClosureTypeRemovedItisCodes {
  closureType: string;
  removedItisCodes: number[];
}

export const getRemovedSimpleClosureTypeItisCodes  = (
  itisCodes: IItisCode[],
): IClosureTypeRemovedItisCodes[] => {
  const closureTypeRemovedItisCodes: IClosureTypeRemovedItisCodes[] = [];
  const codesList = itisCodes.map(itisCode => (itisCode.code));
  Object.keys(closureTypePresetValues)
    .forEach((key) => {
      const newRecord: IClosureTypeRemovedItisCodes = {
        closureType: key,
        removedItisCodes: [],
      };
      closureTypePresetValues[key].itisCodes
        .forEach((cti: number) => {
          const exists = codesList.includes(cti);
          if (exists) return;
          newRecord.removedItisCodes.push(cti);
        });
      closureTypeRemovedItisCodes.push(newRecord);
    });

  return closureTypeRemovedItisCodes.filter(closureType => closureType.removedItisCodes.length > 0);
};
