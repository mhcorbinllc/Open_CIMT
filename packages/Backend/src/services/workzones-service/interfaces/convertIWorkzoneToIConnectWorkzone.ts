import {IWorkzone} from "./IWorkzone";
import {IConnectWorkzone} from "./IConnectWorkzone";
import {IWorkzonesAppSettings} from "./IWorkzonesAppSettings";

export const convertIWorkzoneToIConnectWorkzone = (
  {
    workzone,
    workzoneAppSettings,
  }: {
    workzone: IWorkzone;
    workzoneAppSettings: IWorkzonesAppSettings;
  },
): IConnectWorkzone => {
  return {
    id: workzone.kapschId,
    name: workzone.name,
    speedlimit: Math.round(workzone.speedLimitInMiles),
    isWorkzone: true,
    areWorkersPresent: workzone.workersPresent,
    shoulder: workzone.closedShoulder,
    lane: workzone.closedLane,
    laneWidth: Math.round(workzone.closedLaneWidthInMeters),
    start: dateToIsoString(workzone.start),
    end: dateToIsoString(workzone.end),
    forEver: workzone.forEver,
    itisCodes: workzone.itisCodes,

    // Point
    point: {
      type: "Point",
      coordinates: [
        workzone.point.lng,
        workzone.point.lat,
        workzone.point.alt === undefined
          ? null
          : workzone.point.alt,
      ],
    },

    heading: workzone.heading,

    // Path
    usePath: !!workzone.path.points.length,
    path: {
      type: "LineString",
      coordinates:
        workzone.path.points
          .map(point => ([
            point.lng,
            point.lat,
            point.alt || 0,
          ])),
    },

    // Selection mode: Region
    broadcastRegion: {
      type: "Polygon",
      coordinates: [
        workzone.broadcast.selectionByRegion.region.points
          .map(point => ([
            point.lng,
            point.lat,
            point.alt || 0,
          ])),
      ],
    },

    // Selection mode: Radius
    useRadius: !!workzone.broadcast.selectionByRadius.radiusInMeters,
    radius: workzone.broadcast.selectionByRadius.radiusInMeters,

    // CIMs features options
    serviceChannelStr: workzoneAppSettings.serviceChannel,
    psid: workzoneAppSettings.psid,
    txmode: workzoneAppSettings.txmode,
    interval: workzoneAppSettings.intervalInMs,
    priority: workzoneAppSettings.priority,
    useSignature: workzoneAppSettings.useSignature,
    useEncryption: workzoneAppSettings.useEncryption,
  };
};

const dateToIsoString = (date: Date | string | number): string => {
  return new Date(date).toISOString()
    .split('.')[0] + "Z";
};
