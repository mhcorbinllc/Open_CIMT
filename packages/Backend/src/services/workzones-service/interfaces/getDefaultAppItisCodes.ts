import {IItisCode} from "./IWorkzonesAppSettings";

export const getDefaultAppItisCodes = (): IItisCode[] => {
  return [
    {
      code: 1025,
      label: "Road Construction",
    },
    {
      code: 769,
      label: "Closed to Traffic",
    },
    {
      code: 12545,
      label: "1",
    },
    {
      code: 8196,
      label: "Right Lane",
    },
    {
      code: 8195,
      label: "Left Lane",
    },
    {
      code: 771,
      label: "Closed Ahead",
    },
    {
      code: 7451,
      label: "Merge",
    },
    {
      code: 13580,
      label: "Left",
    },
    {
      code: 13579,
      label: "Right",
    },
    {
      code: 8208,
      label: "Right Shoulder",
    },
    {
      code: 8209,
      label: "Left Shoulder",
    },
    {
      code: 13594,
      label: "Left Turn",
    },
    {
      code: 13599,
      label: "Right Turn",
    },
    {
      code: 268,
      label: "Speed Limit",
    },
    {
      code: 12559,
      label: "15",
    },
    {
      code: 12564,
      label: "20",
    },
    {
      code: 12569,
      label: "25",
    },
    {
      code: 12574,
      label: "30",
    },
    {
      code: 12579,
      label: "35",
    },
    {
      code: 12584,
      label: "40",
    },
    {
      code: 12589,
      label: "45",
    },
    {
      code: 12594,
      label: "50",
    },
    {
      code: 12599,
      label: "55",
    },
    {
      code: 12604,
      label: "60",
    },
    {
      code: 12609,
      label: "65",
    },
    {
      code: 12614,
      label: "70",
    },
    {
      code: 8720,
      label: "MPH",
    },
    {
      code: 2574,
      label: "Height Limit",
    },
    {
      code: 13569,
      label: "Ahead",
    },
    {
      code: 12712,
      label: "168",
    },
    {
      code: 8709,
      label: "Inches",
    },
  ];
};
