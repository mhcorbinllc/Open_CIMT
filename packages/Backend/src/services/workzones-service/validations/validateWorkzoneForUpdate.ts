import {
  IWorkzone,
  EWZLane,
  EWZShoulder,
  EWZBroadcastSelectionMode,
  IGeoPolygon,
  EWZClosureType,
  EWZClosureTypeDeprecated,
} from "../interfaces/IWorkzone";

import {EWZClosureSelectionMode} from "../interfaces/IWorkzonesAppSettings";

import {IValidateDataEngineRules} from "core-library/dist/commonJs/validation-engine";

import {validateDataMethods} from "core-library/dist/commonJs/validation-engine";
import {validateWorkzoneElements} from "../interfaces/validateWorkzoneElements";

// This is very basic validation, we check only security and if the props exist and have correct type.
// We don't validate the logic or if the values are right since we allow the user , this is done on broadcast.

export const validateWorkzoneForUpdate: IValidateDataEngineRules<IWorkzone> = {
  workzoneId: validateDataMethods.hasValueValidationMessage,
  kapschId: validateDataMethods.isNumberValidationMessage,

  active: validateDataMethods.isBooleanValidationMessage,

  name: (name: string) => validateDataMethods.isValidTextValidationMessage(name, 2, 300),
  start: validateDataMethods.isValidDateValueValidationMessage,
  end: validateDataMethods.isValidDateValueValidationMessage,
  forEver: validateDataMethods.isBooleanValidationMessage,
  speedLimitInMiles: validateDataMethods.isIntegerValidationMessage,

  closureSelectionMode: (value: string) => validateDataMethods.isEnumValueValidationMessage(value, EWZClosureSelectionMode),
  closureType: (value: string, workzone) => {
    if (workzone.closureSelectionMode === EWZClosureSelectionMode.CUSTOM) return true;
    return validateDataMethods.isEnumValueValidationMessage(value, EWZClosureType, EWZClosureTypeDeprecated);
  },
  itisCodes: validateDataMethods.isArrayOfNumbersValidationMessage,
  closedLane: (value: string) => validateDataMethods.isEnumValueValidationMessage(value, EWZLane),
  closedShoulder: (value: string) => validateDataMethods.isEnumValueValidationMessage(value, EWZShoulder),

  closedLaneWidthInMeters: validateDataMethods.isNumberValidationMessage,
  workersPresent: validateDataMethods.isBooleanValidationMessage,

  notes: (value: string) => validateDataMethods.isValidTextValidationMessage(value, 0, 10000),

  point: validateWorkzoneElements.isValidGeoPoint,

  path: validateWorkzoneElements.isValidGeoLine,

  heading: validateWorkzoneElements.isValidHeadings,

  broadcast: validateDataMethods.hasValueValidationMessage,
  "broadcast.selectionMode": (value: string) => validateDataMethods.isEnumValueValidationMessage(value, EWZBroadcastSelectionMode),
  "broadcast.selectionByRegion": validateDataMethods.hasValueValidationMessage,
  "broadcast.selectionByRegion.region": (region: IGeoPolygon, workzone) => {
    if (workzone.broadcast.selectionMode === EWZBroadcastSelectionMode.RADIUS) return true;
    return validateWorkzoneElements.isValidPoints(region.points);
  },
  "broadcast.selectionByRadius": validateDataMethods.hasValueValidationMessage,
  "broadcast.selectionByRadius.radiusInMeters": (value: number, workzone) => {
    if (workzone.broadcast.selectionMode === EWZBroadcastSelectionMode.REGION) return true;
    return validateDataMethods.isNumberInRangeValidationMessage(value, 0, 999999);
  },

  createdAt: () => true,  // Ignore them since are overwritten by the db methods anyway
  updatedAt: () => true,
  deletedAt: () => true,
};
