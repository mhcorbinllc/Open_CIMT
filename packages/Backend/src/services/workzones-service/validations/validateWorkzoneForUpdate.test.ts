import {
  EWZBroadcastSelectionMode,
  IWorkzone,
  EWZShoulder,
  EWZLane,
} from "../interfaces/IWorkzone";

import {EWZClosureSelectionMode} from "../interfaces/IWorkzonesAppSettings";

import {getDefaultWorkzone} from "../interfaces/getDefaultWorkzone";
import {validationEngine} from "core-library/dist/commonJs/validation-engine";

import {validateWorkzoneForUpdate} from "./validateWorkzoneForUpdate";
import {EWZClosureTypeDeprecated} from "../interfaces/IWorkzone";

describe('validateWorkzoneForUpdate', () => {
  test('Default CIM', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify({
      ...getDefaultWorkzone(),
      workzoneId: 'ed-98467204',
      name: 'CIM name',
    }));
    expect(validationEngine(workzone, validateWorkzoneForUpdate)).toMatchSnapshot();
  });

  test('Missing workzoneId', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify({
      ...getDefaultWorkzone(),
      name: 'CIM name',
    }));
    expect(validationEngine(workzone, validateWorkzoneForUpdate)).toMatchSnapshot();
  });

  test('Missing radius in meters', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify({
      ...getDefaultWorkzone(),
      workzoneId: 'ed-98467204',
      name: 'CIM name',
    }));
    workzone.broadcast.selectionMode = EWZBroadcastSelectionMode.RADIUS;
    workzone.broadcast.selectionByRadius.radiusInMeters = -1;
    expect(validationEngine(workzone, validateWorkzoneForUpdate)).toMatchSnapshot();
  });

  test('Reduced Speed closure type does not fail closed shoulder/lane', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify({
      ...getDefaultWorkzone(),
      workzoneId: 'ed-98467204',
      name: 'CIM name',
    }));
    workzone.closureSelectionMode = EWZClosureSelectionMode.CUSTOM;
    workzone.closureType = EWZClosureTypeDeprecated.REDUCED_SPEED_AHEAD_30_MPH;
    workzone.closedShoulder = EWZShoulder.NONE;
    workzone.closedLane = EWZLane.NONE;
    expect(validationEngine(workzone, validateWorkzoneForUpdate)).toMatchSnapshot();
  });
});
