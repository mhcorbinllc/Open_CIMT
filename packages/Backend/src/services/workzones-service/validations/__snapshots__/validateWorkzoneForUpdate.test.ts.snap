// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateWorkzoneForUpdate Default CIM 1`] = `
Object {
  "dataValidation": Object {},
  "isValid": true,
  "messages": Array [],
}
`;

exports[`validateWorkzoneForUpdate Missing radius in meters 1`] = `
Object {
  "dataValidation": Object {
    "broadcast.selectionByRadius.radiusInMeters": "Should be a number between 0...999999",
  },
  "isValid": false,
  "messages": Array [
    "Broadcast/selection by radius/radius in meters: Should be a number between 0...999999",
  ],
}
`;

exports[`validateWorkzoneForUpdate Missing workzoneId 1`] = `
Object {
  "dataValidation": Object {
    "workzoneId": "Required",
  },
  "isValid": false,
  "messages": Array [
    "Workzone id: Required",
  ],
}
`;

exports[`validateWorkzoneForUpdate Reduced Speed closure type does not fail closed shoulder/lane 1`] = `
Object {
  "dataValidation": Object {},
  "isValid": true,
  "messages": Array [],
}
`;
